src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the priority queue.
+        /// </summary>
+        public void Clear()
+        {
+            this.nodes.Clear();
+        }
+
+        /// <summary>
+        /// Ensures that the priority queue has the specified capacity
+        /// and resizes its underlying array if necessary.
+        /// </summary>
+        public void EnsureCapacity(int capacity)
+        {
+            if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity));
+            }
+
+            if (capacity <= this.nodes.Count)
+            {
+                return;
+            }
+
+            this.nodes.Capacity = capacity;
+        }
+
+        /// <summary>
+        /// Sets the capacity to the actual number of items in the priority queue,
+        /// if that is less than 90 percent of current capacity.
+        /// </summary>
+        public void TrimExcess()
+        {
+            this.nodes.TrimExcess();
+        }
+
+        /// <summary>
+        /// Removes the node at the specified index.
+        /// </summary>
+        private void Remove(int indexOfNodeToRemove)
+        {
+            // The idea is to replace the specified node by the very last
+            // node and shorten the array by one.
+
+            var lastNodeIndex = this.GetLastNodeIndex();
+            var lastNode = this.nodes[lastNodeIndex];
+            this.nodes.RemoveAt(lastNodeIndex);
+
+            // In case we wanted to remove the node that was the last one,
+            // we are done.
+
+            if (indexOfNodeToRemove == lastNodeIndex)
+            {
+                return;
+            }
+
+            // Our last node was erased from the array and needs to be
+            // inserted again. Of course, we will overwrite the node we
+            // wanted to remove. After that operation, we will need
+            // to restore the heap property (in general).
+
+            var nodeToRemove = this.nodes[indexOfNodeToRemove];
+
+            var relation = this.Comparer.Compare(lastNode.priority, nodeToRemove.priority);
+            this.PutAt(lastNode, indexOfNodeToRemove);
+
+            if (relation < 0)
+            {
+                this.MoveUp(lastNode, indexOfNodeToRemove);
+            }
+            else
+            {
+                this.MoveDown(lastNode, indexOfNodeToRemove);
+            }
+        }
+
+        /// <summary>
+        /// Puts a node at the specified index.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void PutAt((TElement element, TPriority priority) node, int index)
+        {
+            this.nodes[index] = node;
+        }
+
+        /// <summary>
+        /// Gets the index of the last node in the heap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetLastNodeIndex() => this.nodes.Count - 1;
```

Comment:
Nit: inlining `nodes.Count - 1` is pretty self explanatory. I don't think it justifies being factored out to a dedicated method.

Reply:
Does the cost of having a dedicated method outweigh the benefit of: 1) having a readable/verbose expression in the code; 2) deduplication of this expression from its usages into a single place?

Reply:
How is `GetLastNodeIndex()` less readable or verbose than `nodes.Count - 1`? If used with indexers you could simply write `nodes[^1]` to get the last node.

Reply:
You know it's subjective. I can change this if you insist and wouldn't approve the PR without this change. To me it's more readable and verbosity is an advantage here.

Reply:
Yes please. I would say it is pretty standard and we have examples in this codebase where it is being used exactly in that way, e.g. https://github.com/dotnet/runtime/blob/d1d6ab151c693a86736f3dfdefcf09f4bbbdc2bb/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L767 and https://github.com/dotnet/runtime/blob/d1d6ab151c693a86736f3dfdefcf09f4bbbdc2bb/src/libraries/System.Linq/src/System/Linq/Last.cs#L57